packages/loopring_v3/contracts/amm/libamm/AmmAssetManagement.sol

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../lib/ERC20.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/TransferUtil.sol";
+import "./AmmData.sol";
+
+
+/// @title AmmAssetManagement
+library AmmAssetManagement
+{
+    using TransferUtil      for address;
+
+    function deposit(
+        AmmData.State storage S,
+        address               token,
+        uint96                amount
+        )
+        public
+    {
+        if (amount == 0) {
+            return;
+        }
+        uint ethValue = 0;
+        if (token == address(0)) {
+            ethValue = amount;
+        } else {
+            ERC20(token).approve(address(S.exchange.getDepositContract()), amount);
```

Comment:
maybe we can try `S.exchange.deposit` and if it fails, do `approve` then ``S.exchange.deposit` for saving gas.

Reply:
Now with Berlin live I think that may actually end up using more gas than just always approving exactly what is needed. The approval amount is immediately consumed and reset to 0 by the deposit so there are no high storage costs. The token contract needs to be called already for the `transferFrom` so the call to `approve` will also be cheap.